add-chain/addresses.go

```diff
@@ -15,154 +15,89 @@ type AddressData struct {
 	Address string `json:"address"`
 }
 
-var (
-	// Addresses to retrieve from JSON
-	AddressManager                    = "AddressManager"
-	L1CrossDomainMessengerProxy       = "L1CrossDomainMessengerProxy"
-	L1ERC721BridgeProxy               = "L1ERC721BridgeProxy"
-	L1StandardBridgeProxy             = "L1StandardBridgeProxy"
-	OptimismMintableERC20FactoryProxy = "OptimismMintableERC20FactoryProxy"
-	SystemConfigProxy                 = "SystemConfigProxy"
-	OptimismPortalProxy               = "OptimismPortalProxy"
-	ProxyAdmin                        = "ProxyAdmin"
-
-	// Addresses to retrieve from chain
-	SuperchainConfig  = "SuperchainConfig"
-	Guardian          = "Guardian"
-	Challenger        = "Challenger"
-	ProxyAdminOwner   = "ProxyAdminOwner"
-	SystemConfigOwner = "SystemConfigOwner"
-	Proposer          = "Proposer"
-	UnsafeBlockSigner = "UnsafeBlockSigner"
-	BatchSubmitter    = "BatchSubmitter"
-
-	// Non Fault Proof contracts
-	L2OutputOracleProxy = "L2OutputOracleProxy"
-
-	// Fault Proof contracts:
-	AnchorStateRegistryProxy = "AnchorStateRegistryProxy"
-	DelayedWETHProxy         = "DelayedWETHProxy"
-	DisputeGameFactoryProxy  = "DisputeGameFactoryProxy"
-	FaultDisputeGame         = "FaultDisputeGame"
-	MIPS                     = "MIPS"
-	PermissionedDisputeGame  = "PermissionedDisputeGame"
-	PreimageOracle           = "PreimageOracle"
-)
-
-func readAddressesFromChain(addresses map[string]string, l1RpcUrl string, isFaultProofs bool) error {
+func readAddressesFromChain(addresses *superchain.AddressList, l1RpcUrl string, isFaultProofs bool) error {
```

Comment:
There's quite some repetitive boilerplate in this function, maybe it could be done shorter and easier extensible by iterating over a
```go
for _, x := range []struct{
  contract superchain.Address
  call     string
  target   *superchainAddress
  fp       *bool
  allowErr bool
}{
  {
    contract: addresses.OptimismPortalProxy
    call      "superchainConfig()(address)"
    target    &addresses.SuperchainConfig
  },
  // more entries to read
} {
  // call x.call on x.contract and save to *x.target
  // if x.fp non-nil, read entry only if *x.fp == isFaultProofs
  // if allowErr is true, don't return on error, just log
  // could also check if the target address is non-zero before setting, this way this can also be used for the guardian where the first entry could then have allowErr = true, so the second entry is only read if the first failed
  address, err := castCall(...)
}
```
I see that we'd need further logic for the batcher... maybe we also add an optional field `toAddress func(string) Address` and use this one instead of the default `superchain.MustHexToAddress` and then can put the batcher hash extraction code in there.

Reply:
Given urgency to get this in, let's do this in a follow up.

Reply:
Created a new issue to capture this: https://github.com/ethereum-optimism/superchain-registry/issues/427